### ISSUE TITLE
unread messages include already notified ones

### ISSUE THREAD

**Issue by t-animal** [dylex/slack-libpurple#11]
When "retrieve unread history on open" is set, there are messages printed that have already been shown. I.e. "unread history" refers only to unread on slack, but it should include messages in pidgin/client. Maybe messages could be marked as read on slack, when displayed in a window that has focus?

**Comment by dylex** [comment]
That's right, currently this plugin never updates read markers, mainly because it's not entirely clear when to do so. My current thought is to update it every time pidgin's unseen count is set to zero, but this isn't perfect, and there's some work to intercept all the conversation-updated signals. If this sounds reasonable, though, I can work on it.

**Comment by dylex** [comment]
Read markers are now updated on window focus. This seems like a reasonable compromise for how often to update it, and is at least fairly simple. Let me know if this works for you.

**Comment by t-animal** [comment]
Thanks! Will report as soon as I have tested it.

**Comment by t-animal** [comment]
As far as I understand, there is still a small issue: Now, when a message window pops up due to a new message coming in, the read messages are not shown, alright. However the new message is printed twice: once by pidgin in normal colors and once in gray and bold. Presumably the new message is "unread" (which is correct), but imo it should not be included in the "unread messages" backlog, as it is about to be printed by pidgin anyhow.

**Comment by dylex** [comment]
That is a good point. Hm, not exactly sure how to deal with it. You'll first see the real message, and then the history (including that message), and then any further messages (e.g., if 3 is the new message, 3,[1,2,3],4,5...). I thought about deferring the real message until we get the history, which would result in the best display ([1,2],3,4,5,...), but I was worried it would introduce unnecessary delay. I could just remove the final message from history, but then you see messages in a weird order (3,[1,2],4,5...).  Preferences?

**Comment by t-animal** [comment]
My suggestion would be: if there is just one message in the history, discard the history and simply display the message. This covers the use case where a user has pidgin open and gets a message, which will probably cover most messages. If there are more messages in the history I suggest displaying the message first, then "message in context:" or something similar, and then the complete history, including the previously displayed message.

Thanks for your efforts, btw!

**Comment by jaor** [comment]
hi. i am using bitlbee and seeing the original behaviour: the read marker is not updated and i keep retrieving over and over old messages when reconnecting to channels.  it's as if the unseen count is never seen as 0.  i was wondering, is this feature only active when using pidgin as a client? if so, would it be possible to have a setting that says "always update the read marker after retrieving history" or something similar?

thanks!
